libcxx/test/std/numerics/numeric.ops/numeric.iota/ranges.iota.pass.cpp

```diff
@@ -0,0 +1,215 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+// Testing std::ranges::iota
+
+// UNSUPPORTED: c++03, c++11, c++14, c++17, c++20
+
+#include <algorithm>
+#include <array>
+#include <cassert>
+#include <numeric>
+#include <utility>
+
+#include "almost_satisfies_types.h"
+#include "test_iterators.h"
+#include "test_macros.h"
+
+//
+// Testing constraints
+//
+
+// Concepts to check different overloads of std::ranges::iota
+template <class Iter = int*, class Sent = int*, class Value = int>
+concept HasIotaIter = requires(Iter&& iter, Sent&& sent, Value&& val) {
+  std::ranges::iota(std::forward<Iter>(iter), std::forward<Sent>(sent), std::forward<Value>(val));
+};
+
+template <class Range, class Value = int>
+concept HasIotaRange =
+    requires(Range&& range, Value&& val) { std::ranges::iota(std::forward<Range>(range), std::forward<Value>(val)); };
+
+// Test constraints of the iterator/sentinel overload
+// ==================================================
+static_assert(HasIotaIter<int*, int*, int>);
+
+// !input_or_output_iterator<O>
+static_assert(!HasIotaIter<InputIteratorNotInputOrOutputIterator>);
+
+// !sentinel_for<S, O>
+static_assert(!HasIotaIter<int*, SentinelForNotSemiregular>);
+static_assert(!HasIotaIter<int*, SentinelForNotWeaklyEqualityComparableWith>);
+
+// !weakly_incrementable<T>
+static_assert(!HasIotaIter<int*, int*, WeaklyIncrementableNotMovable>);
+
+// !indirectly writable <O, T>
+static_assert(!HasIotaIter<OutputIteratorNotIndirectlyWritable, int*, int>);
+
+// Test constraints for the range overload
+// =======================================
+static_assert(HasIotaRange<UncheckedRange<int*>, int>);
+
+// !weakly_incrementable<T>
+static_assert(!HasIotaRange<UncheckedRange<int*>, WeaklyIncrementableNotMovable>);
+
+// !ranges::output_range<const _Tp&>
+static_assert(!HasIotaRange<UncheckedRange<int*>, OutputIteratorNotIndirectlyWritable>);
```

Comment:
Rather than having concepts, we should move these tests into a `.verify.cpp` file. requires-expressions fail silently, so we don't get signal here on why something like `HasIotaIter<InputIteratorNotInputOrOutputIterator>` is false: only that it is. In a verification test file, we can use Clang to inspect the diagnostic and confirm that `std::ranges::iota(std::forward<Iter>(iter), std::forward<Sent>(sent), std::forward<Value>(val))` fails for the intended reason reason.

Reply:
@cjdb, I'm still new to working on libc++ and wasn't aware of `.verify.cpp` tests so thanks for pointing them out. Are we interested in moving these types of things to `.verify.cpp` files in general? I'm happy to try and do this, I just didn't see other tests doing this (e.g. `std::ranges::generate`) https://github.com/llvm/llvm-project/blob/7ad073a45b411d3752668dff61963d356c98145d/libcxx/test/std/algorithms/alg.modifying.operations/alg.generate/ranges_generate.pass.cpp#L54-L76

Reply:
We should fix those ones too! Here's roughly how I like to write verify files these days:
```cpp
// Each test case becomes its own function, but you can combine iterator/sentinel-based test
// cases with range-based test cases if they're testing the same thing.
void not_input_or_output_iterator(InputIteratorNotInputOrOutputIterator first, int* last, int value)
{
  std::ranges::iota(first, last, value);
  // expected-error@-1{{copy the **error** clang diagnoses here}}
  // expected-note@-2{{copy the **first note** clang diagnoses here}}
  // expected-note@-3{{and so on...}}

  auto r = std::ranges::subrange(first, last);
  std::ranges::iota(r, value);
  // expected-error@-1{{copy the **error** clang diagnoses here}}
  // expected-note@-2{{copy the **first note** clang diagnoses here}}
  // expected-note@-3{{and so on...}}

  std::ranges::iota(std::as_const(r), value);
  // expected-error@-1{{copy the **error** clang diagnoses here}}
  // expected-note@-2{{copy the **first note** clang diagnoses here}}
  // expected-note@-3{{and so on...}}

  std::ranges::iota(std::move(r), value);
  // expected-error@-1{{copy the **error** clang diagnoses here}}
  // expected-note@-2{{copy the **first note** clang diagnoses here}}
  // expected-note@-3{{and so on...}}

  std::ranges::iota(static_cast<decltype(r) const&&>(r), value);
  // expected-error@-1{{copy the **error** clang diagnoses here}}
  // expected-note@-2{{copy the **first note** clang diagnoses here}}
  // expected-note@-3{{and so on...}}
}
```
It's highly likely that we won't want all of the notes, but it might be a good exercise for us to go through them all together (just for one test case: I trust you'll be able to do the others independently).

Reply:
We shouldn't use verify tests when we can test the behaviour without them. The current approach is just fine and mirrors what we do elsewhere.

Reply:
Please expand on how the current tests get around the problem I described in my initial comment.

Reply:
I don't think it's a big problem. You're basically asking to test the test, which doesn't work 99% of the time. This doesn't seem like a super unusual thing to me. OTOH `.verfiy.cpp` tests are extremely fragile and only work for the exact implementation we have. They also don't test that things _actually_ SFINAE away. Anyways, this would require discussion beyond the current patch, since that's how we've done it for all the other ranges algorithms.

Reply:
> I don't think it's a big problem. You're basically asking to test the test, which doesn't work 99% of the time.

It sounds like it doesn't solve the problem at all and I think it is worth our attention. Tests need to be correct _on inspection_, and a static assert isn't offering us that in this case. This limits confidence in our test suite, and I'm not okay with that.

> OTOH `.verfiy.cpp` tests are extremely fragile and only work for the exact implementation we have. 

Diagnostic text doesn't change all that frequently: Clang would have a serious problem if they did (cc @AaronBallman).

How frequently do we (libc++) need to update our existing verification tests?

> They also don't test that things _actually_ SFINAE away.

Are these static assertions actually testing that?

> Anyways, this would require discussion beyond the current patch, since that's how we've done it for all the other ranges algorithms.

We should be committed to doing things right the first time, not to remaining consistent with a status quo in need of change. If that means deviating from what's currently done, that's fine. We can have the discussion afterwards, using this PR as the pilot commit.

I'd really prefer to see this PR landed with maximum confidence in correctness, and also not block @jamesETsmith on a policy discussion.

Reply:
> Diagnostic text doesn't change all that frequently: Clang would have a serious problem if they did (cc @AaronBallman).

We continually update diagnostic text (both in terms of how it reads and in terms of when it fires), so it is a problem that comes up from time to time in Clang. However, the libc++ folks are generally very proactive about helping us come to a good solution in those cases, which limits the blast radius of the problem.

(I don't have a strong opinion on how to resolve the situation here, but my intuition is that testing semantics is more useful than testing diagnostic wording from the perspective of libc++.)

Reply:
> We continually update diagnostic text (both in terms of how it reads and in terms of when it fires), so it is a problem that comes up from time to time in Clang. However, the libc++ folks are generally very proactive about helping us come to a good solution in those cases, which limits the blast radius of the problem.

Thanks for clarifying.

> but my intuition is that testing semantics is more useful than testing diagnostic wording from the perspective of libc++

Right, testing the semantics is important, and knowing that you're testing the right semantics is what the diagnostic wording offers in this case. I'm not as concerned about the specifics of the text, only that it communicates "hey, we're testing what you think we're testing" to the reader.

Reply:
> How frequently do we (libc++) need to update our existing verification tests?

I'm not exactly sure, but there is probably something every 3-4 months on average.

> > Anyways, this would require discussion beyond the current patch, since that's how we've done it for all the other ranges algorithms.
> 
> We should be committed to doing things right the first time, not to remaining consistent with a status quo in need of change. If that means deviating from what's currently done, that's fine. We can have the discussion afterwards, using this PR as the pilot commit.

I absolutely disagree here. "Just put it in sneakily" is _much_ worse than having imperfect code commited. I've seen before that you suggest stuff like `_LIBCPP_ALWAYS_INLINE` or `[[nodiscard]]`. I don't think it's OK for this kind of stuff to be added and then ask for forgiveness. This has significantly eroded my confidence that you give reasonable reviews, especially since people pushed back on these things before.
Just to make it clear: What I _am_ OK with is having a suggestion and asking for other people to give their thoughts on it. If there is no disagreement, fine.

> I'd really prefer to see this PR landed with maximum confidence in correctness, and also not block @jamesETsmith on a policy discussion.

I agree that we should not block this on a policy discussion. That's why I think we should move on with it as-is and not add a `.verify.cpp`. As I said, I don't think it's a good idea and I would block based on this.

Reply:
> > How frequently do we (libc++) need to update our existing verification tests?
> 
> I'm not exactly sure, but there is probably something every 3-4 months on average.

Doesn't seem like it's all that frequent.

> > > Anyways, this would require discussion beyond the current patch, since that's how we've done it for all the other ranges algorithms.
> > 
> > 
> > We should be committed to doing things right the first time, not to remaining consistent with a status quo in need of change. If that means deviating from what's currently done, that's fine. We can have the discussion afterwards, using this PR as the pilot commit.
> 
> I absolutely disagree here. "Just put it in sneakily" is _much_ worse than having imperfect code commited. I've seen before that you suggest stuff like `_LIBCPP_ALWAYS_INLINE` or `[[nodiscard]]`. I don't think it's OK for this kind of stuff to be added and then ask for forgiveness. This has significantly eroded my confidence that you give reasonable reviews, especially since people pushed back on these things before.

I would like to remind you that we have a community [Code of Conduct](https://llvm.org/docs/CodeOfConduct.html), and that it is not okay to accuse folks of being sneaky or unreasonable, just because there is disagreement on technical requests.

With that in mind, it is important to note that incremental changes are a software engineering practice broadly adopted across the industry. This is a practice that LLVM routinely adopts, and it is entirely reasonable to request such changes in a code review. It's fine to disagree with technical suggestions: we can discuss those disagreements and work out what the best path forward is for the whole community.

> Just to make it clear: What I _am_ OK with is having a suggestion and asking for other people to give their thoughts on it. If there is no disagreement, fine.
> 
> > I'd really prefer to see this PR landed with maximum confidence in correctness, and also not block @jamesETsmith on a policy discussion.
> 
> I agree that we should not block this on a policy discussion. That's why I think we should move on with it as-is and not add a `.verify.cpp`. As I said, I don't think it's a good idea and I would block based on this.

We're not inventing a novel kind of test here: there are many similar verification tests in libc++ and adding a set for the range-based algorithms is consistent with those.

Reply:
> > > How frequently do we (libc++) need to update our existing verification tests?
> > 
> > 
> > I'm not exactly sure, but there is probably something every 3-4 months on average.
> 
> Doesn't seem like it's all that frequent.

That's probably the case because we're currently relying on a fraction of the diagnostics clang produces. With every error and note we rely on the frequency will increase. A significant portion of the `.verify.cpp` tests are for deprecated symbols and `static_assert`s currently. Still, having to modify tests every few months seems like a high frequency to me compared to the thousands of tests that have simply existed for the better part of a decade. They tend to not be modified just because we update tools. In the rare cases we have to modify them, it's almost always a bug in the test or in the tool, neither of which are the culprit here.

> > I absolutely disagree here. "Just put it in sneakily" is _much_ worse than having imperfect code commited. I've seen before that you suggest stuff like `_LIBCPP_ALWAYS_INLINE` or `[[nodiscard]]`. I don't think it's OK for this kind of stuff to be added and then ask for forgiveness. This has significantly eroded my confidence that you give reasonable reviews, especially since people pushed back on these things before.
> 
> I would like to remind you that we have a community [Code of Conduct](https://llvm.org/docs/CodeOfConduct.html), and that it is not okay to accuse folks of being sneaky or unreasonable, just because there is disagreement on technical requests.

This isn't a disagreement about a technical question. We disagree on a technical question, but that's not what I'm complaining about here. The disagreement is in how you handle changes of this magnitude. This change isn't very visible to other contributors when doing it in a patch to implement a feature that we've essentially implemented 100 times already. We're in no way breaking new ground in this patch except for this change. This kind of change could easily be done in a patch to ask whether we want to do it, making it much more visible to others.

I'm also not intending to accuse you of being unreasonable. 99% of the work you do is great, but there are a few changes you requested which were very much out-of-place on the review. For that reason I feel I have to check that you're not suggesting something novel to people who don't know the code base very well.

> With that in mind, it is important to note that incremental changes are a software engineering practice broadly adopted across the industry. This is a practice that LLVM routinely adopts, and it is entirely reasonable to request such changes in a code review. It's fine to disagree with technical suggestions: we can discuss those disagreements and work out what the best path forward is for the whole community.

I agree that it's fine to disagree on technical questions. To me it sounds like you don't want to talk to the whole community though. That's the disagreement I have.

> We're not inventing a novel kind of test here: there are many similar verification tests in libc++ and adding a set for the range-based algorithms is consistent with those.

There may be some, but if there is a reasonable way to test SFINAE we generally do that without `.verify.cpp` tests. These tests are primarily used to used for things that should be ill-formed (even in a SFINAE context) and for warnings, neither of which are checked for here.

Reply:
> Right, testing the semantics is important, and knowing that you're testing the right semantics is what the diagnostic wording offers in this case. I'm not as concerned about the specifics of the text, only that it communicates "hey, we're testing what you think we're testing" to the reader.

Sort of yes, sort of no. In order to test that you're getting the right semantics, you need to test against specific diagnostic wording otherwise you run the risk of passing the test because of an unrelated but similarly worded diagnostic. Once you're testing against sufficiently specific diagnostic wording, it becomes more of a maintenance burden should Clang update that diagnostic wording (which is something we do somewhat frequently, but we have 1000s of warnings, so it's a statistics game as to whether any given diagnostic is likely to be hit). So I think it's easier for everyone to maintain if the test can test behavior rather than diagnostic wording whenever possible even if the tests are more complex, but there are certainly cases where it's not possible (e.g., a deprecation warning can really only be tested by looking for the diagnostic to be emitted).

Reply:
> > > > How frequently do we (libc++) need to update our existing verification tests?
> > > 
> > > 
> > > I'm not exactly sure, but there is probably something every 3-4 months on average.
> > 
> > 
> > Doesn't seem like it's all that frequent.
> 
> That's probably the case because we're currently relying on a fraction of the diagnostics clang produces. With every error and note we rely on the frequency will increase. A significant portion of the `.verify.cpp` tests are for deprecated symbols and `static_assert`s currently. Still, having to modify tests every few months seems like a high frequency to me compared to the thousands of tests that have simply existed for the better part of a decade. They tend to not be modified just because we update tools. In the rare cases we have to modify them, it's almost always a bug in the test or in the tool, neither of which are the culprit here.
> 
> > > I absolutely disagree here. "Just put it in sneakily" is _much_ worse than having imperfect code commited. I've seen before that you suggest stuff like `_LIBCPP_ALWAYS_INLINE` or `[[nodiscard]]`. I don't think it's OK for this kind of stuff to be added and then ask for forgiveness. This has significantly eroded my confidence that you give reasonable reviews, especially since people pushed back on these things before.
> > 
> > 
> > I would like to remind you that we have a community [Code of Conduct](https://llvm.org/docs/CodeOfConduct.html), and that it is not okay to accuse folks of being sneaky or unreasonable, just because there is disagreement on technical requests.
> 
> This isn't a disagreement about a technical question. We disagree on a technical question, but that's not what I'm complaining about here. The disagreement is in how you handle changes of this magnitude. This change isn't very visible to other contributors when doing it in a patch to implement a feature that we've essentially implemented 100 times already. We're in no way breaking new ground in this patch except for this change. This kind of change could easily be done in a patch to ask whether we want to do it, making it much more visible to others.

This request is for long-term maintainability, and confidence in the code. I'm not asking for drastic changes to be made here: I'm asking for a single test to either be rewritten in a way that captures the missing functionality, or to add a new test that captures it if the existing test is fulfilling a role that the rewrite does not capture. Just because we have done something one way a hundred times is not a good argument for us to continue doing it that way with new code. Software engineering is based on incremental change, and that is what I am asking for.

As for visibility, I'm not sure what you're referring to: I have publicly requested the change in a public forum. That is rather visible.

> I'm also not intending to accuse you of being unreasonable. 99% of the work you do is great, but there are a few changes you requested which were very much out-of-place on the review. For that reason I feel I have to check that you're not suggesting something novel to people who don't know the code base very well.
> 
> > With that in mind, it is important to note that incremental changes are a software engineering practice broadly adopted across the industry. This is a practice that LLVM routinely adopts, and it is entirely reasonable to request such changes in a code review. It's fine to disagree with technical suggestions: we can discuss those disagreements and work out what the best path forward is for the whole community.
> 
> I agree that it's fine to disagree on technical questions. To me it sounds like you don't want to talk to the whole community though. That's the disagreement I have.

At what point have I expressed a desire to avoid discussion about this? Just a few comments up, I said that we should have a discussion about the material, using the changes from this PR as information for future work.

If you aren't complaining about something technical, then why is it being discussed in a pull request?
More appropriate avenues of discussion for such a topic include:

  * directly messaging me to understand what I'm doing and why, so that we can collaborate on a path forward
  * asking in the libc++ group chat if you want a more visible discussion
  * escalating to the Code of Conduct committee, if you believe my behaviour is a violation of the rules put in place

It feels like a lot of time has been spent telling me what my intentions are; no time has been spent asking questions to confirm that these suspicions are grounded, or even to gather information on understanding why there are conflicting technical viewpoints, both of which are critical for collaboration. I would appreciate it if more time was spent doing the latter before concluding that a person is behaving in an antagonistic fashion towards the libc++ community.

Reply:
> This request is for long-term maintainability, and confidence in the code. I'm not asking for drastic changes to be made here: I'm asking for a single test to either be rewritten in a way that captures the missing functionality, or to add a new test that captures it if the existing test is fulfilling a role that the rewrite does not capture. Just because we have done something one way a hundred times is not a good argument for us to continue doing it that way with new code. Software engineering is based on incremental change, and that is what I am asking for.

Because we did it a hundred times, it's a good idea to not change it in the 101st iteration, but instead separately. I don't think it's a simple extension of the test and I think I've made that pretty clear.

> As for visibility, I'm not sure what you're referring to: I have publicly requested the change in a public forum. That is rather visible.

People tend to not look through every PR they see. They simply don't have the time for that. If they see the 100th time that a ranges algorithm is implemented, they don't expect a significant change in how it's done. What I'm asking for is a separate PR which clearly states in the title what change you're proposing here. That tends to spark people's interest, since it's not the 100th time they see it.

> > I'm also not intending to accuse you of being unreasonable. 99% of the work you do is great, but there are a few changes you requested which were very much out-of-place on the review. For that reason I feel I have to check that you're not suggesting something novel to people who don't know the code base very well.
> > > With that in mind, it is important to note that incremental changes are a software engineering practice broadly adopted across the industry. This is a practice that LLVM routinely adopts, and it is entirely reasonable to request such changes in a code review. It's fine to disagree with technical suggestions: we can discuss those disagreements and work out what the best path forward is for the whole community.
> > 
> > 
> > I agree that it's fine to disagree on technical questions. To me it sounds like you don't want to talk to the whole community though. That's the disagreement I have.
> 
> At what point have I expressed a desire to avoid discussion about this? Just a few comments up, I said that we should have a discussion about the material, using the changes from this PR as information for future work.

That discussion should be done _before_ the change, not after. That's what I meant with "asking for forgiveness later".

> If you aren't complaining about something technical, then why is it being discussed in a pull request? More appropriate avenues of discussion for such a topic include:
> 
>     * directly messaging me to understand what I'm doing and why, so that we can collaborate on a path forward
> 
>     * asking in the libc++ group chat if you want a more visible discussion
> 
>     * escalating to the Code of Conduct committee, if you believe my behaviour is a violation of the rules put in place

If you think that's more productive, feel free to send me a PR so we can set up a meeting to discuss this further.
 
> It feels like a lot of time has been spent telling me what my intentions are; no time has been spent asking questions to confirm that these suspicions are grounded, or even to gather information on understanding why there are conflicting technical viewpoints, both of which are critical for collaboration. I would appreciate it if more time was spent doing the latter before concluding that a person is behaving in an antagonistic fashion towards the libc++ community.

I can't find a single place where I told you what your intentions are. I simply stated how I perceived some of your actions.

Reply:
@cjdb, @philnik777, and @AaronBallman, thanks again for taking the time to look over my PR. Were you able to reach a consensus offline about how to move forward on this?

Reply:
Still discussing, the ball is in my court currently.

Reply:
Hey @jamesETsmith, sorry for the delay. Let's progress with the status quo in this patch, and I'll keep thinking about testing strategy.